### ISSUE TITLE
Lock down mouse actions

### ISSUE THREAD

**Issue by JuliusSweetland** [OptiKey/OptiKey#421]
Ability to lock down a mouse action and have it repeat over and over

**Comment by JuliusSweetland** [comment]
An example of software which supports the repeat click of the mouse: http://iris.xcessity.at/

**Comment by kmcnaught** [comment]
Why was this one closed? I was about to make exactly this request... :)  I think it would tie in really nicely with a new release containing a variety of mouse improvements...

**Comment by JuliusSweetland** [comment]
@kmcnaught I did an agressive clean up. You raised this one as sitll pertient, so back from the grave it comes...

**Comment by AdamRoden** [comment]
Xcessity... The software that gave me the drive and inspiration to get involved in improving Optikey. 
Can't repetitive clicking be accomplished via a loop command in a key on a dynamic keyboard?

**Comment by kmcnaught** [comment]
@AdamRoden it might be possible, but I'd want to make sure it was functional in the main mouse keyboard, not just available for advanced users with dynamic keyboards. This is the default behaviour in pretty much any basic cursor control software on the assistive market, so users coming from that background will expect it to be easily available. I think it particularly makes a difference when someone needs long dwell movements, so completing a task such as "navigate Windows Explorer to find the file I want" becomes quite tedious if every click requires a separate action in Optikey.

**Comment by JuliusSweetland** [comment]
@kmcnaught @AdamRoden It could be achieved with a MC setting which makes the default behaviour of some mouse actions (maybe only left click, but we could add configuration for middle and right if appropriate) that they repeat until turned off.

Actually, what about another key on the expanded keyboard which allows you to turn on, or lock on, repeated left clicks. This gives users full control to enable/disable this. We could back up the button state into a setting so it is retained across restarts.

Would we ever want this for any key other than left click?

Thoughts?

**Comment by kmcnaught** [comment]
I was imagining you'd just have the left click key (and others if appropriate?) be lockable, so if you click once they perform one action, but if you click twice to lock down they keep repeating until switched off. This would be simpler from a UX perspective and consistent with existing Optikey UX, but maybe it's non-trivial on the implementation side? One complication is that you'd have to disable the dwell-at-point functionality while over the keyboard, so that it doesn't interfere with the second dwell on the mouse key. 

If the above isn't feasible then having a lock toggle on the expanded keyboard could work. That would be nicer (more discoverable) than a setting in MC. 

I hadn't considered whether this applies further than left click - probably not that common? If the individual key(s) itself was lockable then it would make sense to allow the same functionality on other mouse actions - there may be some specific situations where you want lots of double clicks for instance, and the rest of the time you just don't ever lock the key down. If it's done via a toggle then maybe you'd only apply to left click? e.g. while carrying out your task you might want lots of left clicks with the occasional (single) right click in between. In the latter case you could have an MC setting to extend the behaviour, if anyone requests it.

**Comment by JuliusSweetland** [comment]
The UX problem with locking the left click key itself (for example) is how
do you know what the user intends to do after the first key press - do they
want to begin a click event, or press the click key again to lock it?

**Comment by kmcnaught** [comment]
Yeah, makes sense. I don't _think_ there's any reason a user would be using a magnified mouse click on the main Optikey keyboard, so in theory you could treat the cases separately, but I understand it might give some tricky implementation and/or UX around the edge of the keyboard. 

Your suggestion of a single "lock multi-click" toggle key skirts around the issue and works perfectly well if a user generally wants one of the two possible styles of clicking, but is a bit less flexible if you like to mix and match (it's nice to be able to do as much as possible without expanding the keyboard). I don't know how often that would be the case in practise. Certainly having the ability to choose _either_ multiclick _or_ single click is already a big improvement, so don't take my preferences too seriously ;)

**Comment by JuliusSweetland** [comment]
Agree with all that.

The only thing I can think for why a user might want to click on Optikey
itself is a right click to launch the context menu.

**Comment by kmcnaught** [comment]
fwiw I just had another request for this from a user.

**Comment by jvanewijk** [comment]
> fwiw I just had another request for this from a user.

Yes, that would be from me I guess. I'm not a developer, neither a user of Optikey myself. But I'm an occupational therapist and have a lot of experience with people using eyegaze. I follow Optikey for some years now and I'm so happy to see how it develops! 

About this issue: it would be very helpful to users if it would be possible to lock the left click. For example to speed up clicking. If you are on a website and have to click a lot of links it can take a lot of time to activate the button each time. 
For some users with cognitive impairment it is easier to have the click always on: you look you click. Easy. 

And then about making other keys lockable: I know some eyegaze users love to paint. They did this with a locked single left click and/or a locked mouse drag. Since the newer Tobii software (Control it is called) I read a lot of complaints that this wasn't possible any more. Eventually they added the feature, but now you have the lock on for example the mouse drag and you can't use another mouse click until you unlock the drag. So if you are painting (with the locked drag function) and you want to change color (you will need a single left click) you have to unlock mouse drag, choose single left click, choose another color, lock mouse drag again. That's quite a lot of steps. I think it would be better if you would be able to go from locked mouse drag to single left click without first unlocking mouse drag. Does that make sense? 

I know a lot of people are complaining about this new software from Tobii. I think Optikey is already an alternative for them. And I think it will be even better if a lock key will be available some time. 
As I said, I can't help you with programming it. But I am available if you need any testing or maybe something else. Just let me know! 
Jikke

**Comment by richardchui** [comment]
> > fwiw I just had another request for this from a user.
> 
> Yes, that would be from me I guess. I'm not a developer, neither a user of Optikey myself. But I'm an occupational therapist and have a lot of experience with people using eyegaze. I follow Optikey for some years now and I'm so happy to see how it develops!
> 
> About this issue: it would be very helpful to users if it would be possible to lock the left click. For example to speed up clicking. If you are on a website and have to click a lot of links it can take a lot of time to activate the button each time. For some users with cognitive impairment it is easier to have the click always on: you look you click. Easy.
> 
> And then about making other keys lockable: I know some eyegaze users love to paint. They did this with a locked single left click and/or a locked mouse drag. Since the newer Tobii software (Control it is called) I read a lot of complaints that this wasn't possible any more. Eventually they added the feature, but now you have the lock on for example the mouse drag and you can't use another mouse click until you unlock the drag. So if you are painting (with the locked drag function) and you want to change color (you will need a single left click) you have to unlock mouse drag, choose single left click, choose another color, lock mouse drag again. That's quite a lot of steps. I think it would be better if you would be able to go from locked mouse drag to single left click without first unlocking mouse drag. Does that make sense?
> 
> I know a lot of people are complaining about this new software from Tobii. I think Optikey is already an alternative for them. And I think it will be even better if a lock key will be available some time. As I said, I can't help you with programming it. But I am available if you need any testing or maybe something else. Just let me know! Jikke

Just to echo, this would be a huge time saver when using AAC applications.

**Comment by kmcnaught** [comment]
recent discussion over at #744
